File: save_dynamic_node_info.py
import os
import subprocess

from tqdm import tqdm  # type: ignore

from utils import NodeNameConverter


def run_command(command):
    result = subprocess.run(command, stdout=subprocess.PIPE, text=True, shell=True)
    return result.stdout


OUTPUT_DIR = 'dynamic_node_info'
SKIP_NODES = ['caret_', 'launch_ros_', 'rviz2', 'rosbag2_player', 'transform_listener_impl_']


def main():
    node_list = run_command('ros2 node list').strip().split('\n')
    for node_name in tqdm(node_list):
        if any(skip_node in node_name for skip_node in SKIP_NODES):
            continue

        info_str = run_command(f'ros2 node info {node_name}')
        info_lines = info_str.strip().split('\n')[1:]
        info_text = '\n'.join(info_lines)

        node_file_name = NodeNameConverter.to_file_name(node_name)
        os.makedirs(f'{OUTPUT_DIR}', exist_ok=True)
        with open(f'{OUTPUT_DIR}/{node_file_name}.yaml', 'w', ) as f:
            f.write(info_text)


if __name__ == "__main__":
    main()

File: parse_node_info/src/bin/save_dynamic_node_info.rs
use parse_node_info::utils::{create_progress_bar, NodeNameConverter};
use std::fs::{self, File};
use std::io::Write;
use std::process::Command;

const OUTPUT_DIR: &str = "dynamic_node_info";
const SKIP_NODES: [&str; 13] = [
    "caret_",
    "launch_ros_",
    "rviz2",
    "rosbag2_player",
    "transform_listener_impl_",
    "container",
    "/system/component_state_monitor/component", // not executable
    "default_ad_api",                            // not working in rosbag replay simulation
    "autoware_api/external/rtc_controller/node", // not working in rosbag replay simulation
    "robot_state_publisher",
    "map_hash_generator",
    "/perception/traffic_light_recognition/traffic_light_arbiter/arbiter", // not executable
    "aggregator_node",
];

pub fn run_command(command: &str) -> String {
    let output = Command::new("sh")
        .arg("-c")
        .arg(command)
        .output()
        .expect("failed to execute command");
    String::from_utf8(output.stdout).unwrap()
}

pub fn main() {
    let node_list_str = run_command("ros2 node list");
    let node_list = node_list_str.trim().split('\n');
    let pb = create_progress_bar(node_list.clone().count() as i32);
    for ros_node_name in node_list {
        pb.inc(1);
        if SKIP_NODES
            .iter()
            .any(|&skip_node| ros_node_name.contains(skip_node))
        {
            continue;
        }

        let info_str = run_command(&format!("ros2 node info {}", ros_node_name));
        let info_lines: Vec<&str> = info_str.trim().split('\n').collect();
        let info_text = info_lines[1..].join("\n");

        fs::create_dir_all(OUTPUT_DIR).unwrap();
        let mut file = File::create(format!(
            "{}/{}.yaml",
            OUTPUT_DIR,
            NodeNameConverter::to_file_name(ros_node_name)
        ))
        .unwrap();
        file.write_all(info_text.as_bytes()).unwrap();
    }
}
